### DISCUSSION TITLE
Why Is Restore Option Missing Entirely?

### DISCUSSION THREAD

**Discussion by Jay-study-nildana** [microsoft/azuredatastudio#14814] [Q&A]
I hope its okay to link stack overflow here. It has images and everything.  

https://stackoverflow.com/questions/66723032/azure-data-studio-why-is-restore-option-missing-entirely

I simply dont get the "Restore Option" on Azure Data Studio. What am I missing?

**Comment by Charles-Gagnon** [comment]
Currently the backup/restore feature does not support Azure SQL DBs. You can configure backup retention and restore from the Azure portal : https://docs.microsoft.com/en-us/azure/azure-sql/database/recovery-using-backups

**Reply by Jay-study-nildana** [reply]
@Charles-Gagnon 

That's not really my question. It looks like the option is not currently available on Mac. 

I ended up using SQL Management Studio on my windows machine to get through the database restore. thank you :)

**Reply by Charles-Gagnon** [reply]
What do you mean it's not available on Mac? It should be available there as long as you're not connected to an Azure or SQL On-Demand server.

I was basing my response off of the linked stack overflow question - which showed an Azure database being connected to. If you have a different situation could you please attach an image of the dashboard as you see it?

**Reply by Jay-study-nildana** [reply]
Wait a minute. 

" It should be available there as long as you're not connected to an Azure or SQL On-Demand server."

I don't understand this limitation. Perhaps I am misreading the purpose of Azure SQL Studio. is it only to run queries? I thought it was being positioned as a replacement of SQL Management Studio. 

My scenario is

1. I have a .bak file/.bacpac file.
2. Upload it to Azure SQL db. 

This is something I can do on SQL Management Studio (on my windows computer). I was experimenting with SQL on my Mac, and I thought, wait a minute, perhaps I can use this new Azure Studio app to accomplish some basic Management Studio tasks. 

I am just tired of lugging my windows laptop everywhere, just to restore a .bak .bacpac file.

**Reply by Jay-study-nildana** [reply]
Further,

I am not using an On-Demand Server.

I am using the basic SQL instance database.

As per this tutorial link, https://docs.microsoft.com/en-us/sql/azure-data-studio/tutorial-backup-restore-sql-server?view=sql-server-ver15#restore-a-database-from-a-backup-file,

I am supposed to see something that looks like this. 

![tasks-restore](https://user-images.githubusercontent.com/14297315/112566630-6119e600-8e05-11eb-9dd5-ec8bdb53b609.png)

However, here is a fresh screenshot I took this morning (similar to the image I have put on my stack overflow question). As per the tutorial, I have selected the server in question before waiting for the indicated 'Restore' options to show up. 

<img width="1412" alt="Screenshot 2021-03-26 at 7 28 31 AM" src="https://user-images.githubusercontent.com/14297315/112566680-73941f80-8e05-11eb-9685-bf1f193035c6.png">

Either I am missing something that is really obvious, or, the feature is indeed missing. 

Version details, if it helps.

    Version: 1.27.0
    Commit: 765c7064d46253d88393e0e1ff78874b6fe8fa64
    Date: 2021-03-16T21:18:54.187Z
    VS Code: 1.51.0
    Electron: 9.4.3
    Chrome: 83.0.4103.122
    Node.js: 12.14.1
    V8: 8.3.110.13-electron.0
    OS: Darwin x64 20.2.0

Also, I noticed that, I dont get the option to backup an existing database either. This link, https://docs.microsoft.com/en-us/sql/azure-data-studio/tutorial-backup-restore-sql-server?view=sql-server-ver15#back-up-a-database, says I should see some nice visualisations plus those 4 icons, which includes, backup. 

I dont get that either. Here is a photo with a database selected.

<img width="1412" alt="Screenshot 2021-03-26 at 7 37 23 AM" src="https://user-images.githubusercontent.com/14297315/112567035-1d73ac00-8e06-11eb-89cc-59accd9beb80.png">

**Reply by Charles-Gagnon** [reply]
The server name is *.database.windows.net, which implies it's hosted in Azure. Could you run @@SERVERVERSION on the server to verify?

Restoring a database from a BAK file is not supported for Azure SQL servers - this is a limitation of the engine. You can use a BACPAC to import a database though https://docs.microsoft.com/en-us/azure/azure-sql/database/database-import?tabs=azure-powershell, which you can use the Dacpac extension in ADS for.

**Reply by Jay-study-nildana** [reply]
@Charles-Gagnon 

I apologise and I dont mean to be rude. My question seems to be continuously misunderstood, consistently.

My issue (and I thought its obvious from the pictures) is that I am not getting the options as per the tutorial linked images.  

My question is not about this - "Restoring a database from a BAK file is not supported for Azure SQL servers ". Thats really irrelevant right now. 

Really, it's okay. a client of mine wanted to do SQL stuff on Mac. I suggested Azure Studio, but now, both client and me have reverted to SQL Management Studio on Windows. 

Let's table this for later. I am in the middle of the project and I cannot troubleshoot and follow up with steps right now. 

Thank you for trying to help me, but really, Azure Studio is just not ready for production use, is the conclusion I draw from our conversation. Me and My client will return to Azure Studio a few months later :)

For now, we are happy with SQL Management Studio.

Thanks again.

**Reply by Charles-Gagnon** [reply]
I apologies for the difficulties here @Jay-study-nildana - there does seem to be some miscommunication happening. I appreciate your patience, and if possible would love to help figure out whether we have a gap in our offerings for ADS that isn't fully documented. 

The documentation doesn't seem to mention that backup/restore for an Azure DB is not currently possible and so I will definitely work to get that updated. But I also want to make sure we're not missing any other scenarios that you might have ran in to.

In your screenshot the dashboard shows that the database you're connected to is indeed an Azure SQL Instance : 

![image](https://user-images.githubusercontent.com/28519865/112864336-9411e300-906c-11eb-9c91-3bef9f6102cb.png)

If that is the case then doing backup/restore from the GUI in both SSMS/ADS is impossible - the option should not be showing up in either tool. 

Your response seems to indicate that you're able to successfully use SSMS to initiate the backup/restore though - is this the case? If so running `SELECT @@VERSION` and copying that here would help me make sure that my assumptions above are correct.

**Reply by dzsquared** [reply]
> Wait a minute.
> 
> " It should be available there as long as you're not connected to an Azure or SQL On-Demand server."
> 
> I don't understand this limitation. Perhaps I am misreading the purpose of Azure SQL Studio. is it only to run queries? I thought it was being positioned as a replacement of SQL Management Studio.
> 
> My scenario is
> 
> 1. I have a .bak file/.bacpac file.
> 2. Upload it to Azure SQL db.
> 
> This is something I can do on SQL Management Studio (on my windows computer). I was experimenting with SQL on my Mac, and I thought, wait a minute, perhaps I can use this new Azure Studio app to accomplish some basic Management Studio tasks.
> 
> I am just tired of lugging my windows laptop everywhere, just to restore a .bak .bacpac file.


@Jay-study-nildana - thanks for walking through these scenarios.  I wanted to peel back 1 layer that is combined above - bak and bacpac files.

bak files are a backup type and format specific to SQL Server and are not supported in Azure SQL DB.  The dashboard interface for backup/restore uses this bak file type.

bacpac files are supported in both SQL Server and Azure SQL DB, however to work with these files in Azure Data Studio you will need to install the extension "SQL Server Dacpac".  <-- you want to install this to restore your Azure SQL DB from the bacpac file you have

**Reply by Jay-study-nildana** [reply]
@Charles-Gagnon 

**Part One**

There is no need for apologies. We are all just trying to get work done. It's all cool for me. It's not like any of this is affecting work. My client and I are big fans of Management Studio. Old is gold, as the saying goes :P 

But, like any responsible developer, I like to try out a few new things. 

I think the issue really boils down to "there does seem to be some miscommunication happening"

The official documentation pages try to sell you and make you believe that Azure Studio is primed to take over some basic tasks currently being done by Management Studio. Perhaps I am reading too much into the documentation but as a user of .NET for more than 20 years, having lived through the MSDN days of documentation, I think I can read and understand documentation just fine. 

I would like to think restoring a simple .bacpac file (Let's forget about .bak for now. It's fine) to the standard azure sql service should be like the first thing that a 'supposed' replacement to management studio should be able to do. 

I think, the messaging is not correct. It's not pointing out the obvious and the images clearly dont make any sense. For example, why put a cropped photo of just the 'restore' portion of the total image. That's never a good idea. Why not put both the full image of the application (so we can see all the options that are selected) and also the cropped image?

[I understand the documentation is open source. I got one or two commits on Microsoft Docs myself, where previously, I found some issues and my pull requests were accepted. I was hoping to updated this image myself, add necessary clarity until, I ran into the very issue we are discussing]

I prepare documentation for my clients, and I put an image like that, I would get dinged for sure. I would get a lot of complaints.

**Part Two**

Actually, "doing backup/restore from the GUI in both SSMS/ADS is impossible" - This is absolutely correct. 

On Management Studio, 

- Step One: I had to restore the .bak to a local database first. 
- Step Two: Then, sync it with the online server using the 'publish' to azure option. 

I suppose, this is what I was hoping for in Azure Data Studio. Perhaps, I should not be expecting this. 

However, as I have mentioned before, my issue is not if I can restore .bak or .backpac files. We haven't reached that stage yet. 

My issue remains, the buttons to do the restore (.bak or .bacpac or .xyzsomethingsomething), are simply missing from the tool entirely. 

And yes, I am okay if the issue is not resolved or no solution exists. It's totally cool folks. 

**Lastly,**

I just want to add that, my client has removed Azure Data Studio from her computer. However, I am now using it on both of my Macs on a regular basis. Albeit, with the understanding that, back up and restore, I will still need to use my old trusty Dell, he he :) 

The Data Studio super fast and super lightweight. and it look's cool. I frequently have to engage with young college kids on training sessions. They look at Management Studio and they will go, "Dude, what is this old thing?". Kids you know, thats how they are. But now, Azure Data Studio. It's good. The kids like it a lot. They still complain about having to write SQL queries, but at least, they are not complaining about the looks of the tool. 

With all due respect, Management Studio, is from the old days days. Just like Visual Studio, it's large and resource heavy, but very powerful, of course. 

It's all good from my side. Backup and Restore is something I do once in a few months. I really dont mind if that feature never comes to Azure Studio.

**Reply by Jay-study-nildana** [reply]
@dzsquared 

"you will need to install the extension "SQL Server Dacpac""

I will look into this the first chance I get off from work this week. 

I will update with my results. Thank you.
